### DISCUSSION TITLE
Workflow Advice - jumping to and from certain PDF pages

### DISCUSSION THREAD

**Discussion by ryanwwest** [koreader/koreader#10881] [Q&A] [answered]
I use Koreader to read a _lot_ of 200+ page PDFs and oftentimes the beginning of a document has an 8-page alphabetic table of acronyms that I must continuously refer back to. To quickly do this, my workflow is to create a bookmark at the center page of this acronym table (or wherever I want to refer back to) and then turn on 'Bookmark browsing mode' which changes the left/right swipe actions from previous/next page to previous/next bookmark (and highlight which I think is also considered a bookmark). 

As long as I have just one bookmark, a swipe from left to right immediately takes me to the bookmarked page (awesome!), then I will often tap up to 3-4 pages forward or backward to reach different parts of the acronym table. Then, to return to the current reading position in the entire document, I mapped the Settings->Navigation->Back in reader setting to 'previous read page' which lets me press the device's back button up to 4-5 times to go back page by page until it jumps me to the current reading position. (I had hoped that the alternate setting 'Go to previous location' would work, but it seems to forget the current reading position that way and takes me to some older page).

There are two issues:
1. I'd like to find a way to jump immediately back to the current reading page with one action, not 4-5 actions like above.
2. I can't highlight with this method, as if I do, they are also considered bookmarks and I have to swipe through all bookmarks to get to the acronym table at the start of the PDF.

I was wondering if there are any better solutions for improving this workflow? I wouldn't want to bookmark the current page I'm reading since it's only temporarily important, unlike the acronym table.

**Comment by hius07** [comment]
It might be to assign several actions to a gesture, for example:
One finger swipe Bottom edge right: Toggle bookmark + Previous bookmark (you are in the acronyms section)
One finger swipe Bottom edge left: Next bookmark + Toggle bookmark (back on the page)
No need to switch to the flipping mode.

**Reply by ryanwwest** [reply]
Thanks, that seems like a way to solve (1). It seems like the order in which you select multiple actions for a given gesture matters (though it isn't indicated), so it only works in the order you describe. It also seems to be working with this, which doesn't accidentally mess up bookmarks if I misswipe:
- One finger swipe Bottom edge right: Previous bookmark (you are in the acronyms section)
- One finger swipe Bottom edge left: Previous Location (back on the page - even if I stepped a few pages left/right in the acronyms)

Of course, (2) is still unsolved. I was hoping that I could do:
- One finger swipe Bottom edge right" to three actions: Toggle bookmark -> First page -> Next bookmark 
- One finger swipe Bottom edge left: Latest bookmark -> Toggle bookmark

and be okay with not adding highlights before the table (since it's usually near the start of the PDF) and then highlights/additional bookmarks could work, but it looks like only the first 'page movement' action can occur despite all 3 actions being assigned in an attempted order. It would be cool if the gestures could work like 'macros' with many repeating actions or you could define your own macro actions and assign them to gestures, but that's probably too niche.

I'm sure you've heard of the Sioyek academic PDF reader - its smart jump and portal features are pretty neat: https://sioyek.info/. I know KOreader is a different type of product and am not looking for feature parity, but it's still cool!

**Reply by hius07** [reply]
> the order in which you select multiple actions for a given gesture matters

There is "Sort" menu item in the gesture actions menu.

**Reply by ryanwwest** [reply]
Thanks for pointing that out. So it seems like (2) isn't feasible (i.e., you can't assign Toggle bookmark -> First page -> Next bookmark and have it work properly currently, assuming that's the only solution)?

**Reply by hius07** [reply]
Making bookmarks could be possible in your workflow, but it needs to add (code) one action: Go to the first bookmark (which would be in the acronym section).
One finger swipe Bottom edge right: Toggle bookmark + Go to the first bookmark
One finger swipe Bottom edge left: Go to the latest bookmark (this action already exists) + Toggle bookmark

**Reply by hius07** [reply]
Or, instead of "Go to the first bookmark" it can be "Go to page #" (doesn't exist either).

**Reply by ryanwwest** [reply]
I can think of 4 solutions:

A. Your first suggestion (implement new action to go to first bookmark (maybe less flexible and potentially obviated by (C))
B. Your second suggestion (implement new action to go to a specific page defined just there) (more flexible, i.e. creates a higher-priority 'bookmark' that circumvents highlight bookmark issues; but maybe harder to implement selecting page # in UI)
C. Allow 'macros' in gesture action order (i.e. allow any number of actions to occur in order including First page -> Next bookmark; or might be better as new custom 'actions' defining a sequence of actions themselves, not just bound to gestures, and thus increase implementation complexity)
D. Let Bookmarks Browsing Mode and 'jump to next/previous/latest bookmark' optionally consider only page-bookmarks or only highlight-bookmarks in addition to both.

(B) seems the best here, though (C) could open up some serious power user options and could be a future idea.

**Reply by ryanwwest** [reply]
Thinking more about (B), an action that jumps to a specific page number would only be useful per-document and require modifying the gesture action each time you read a new one since a different page might be useful somewhere else. I guess that makes (A) more flexible in that regard, though the jump-to-page or higher-priority bookmark idea is very nice in theory...

**Reply by hius07** [reply, marked as answer]
> custom 'actions' defining a sequence of actions

They are called Profiles (in the Tools menu).
Create a profile, add any actions to it, check Show in action list, the profile will appear in General section of the Taps and gestures menu, and can be assigned to any gesture.

**Reply by ryanwwest** [reply]
This is incredible, it's already implemented! Thank you for educating me. I don't think I could use the same gesture (like tap lower left corner) to 'toggle' between a profile that jumps to the first bookmark and jumping back to the 'previous location' but I can accomplish it in two gestures while using additional bookmarks/highlights after the first, which solves my problem.

An idea of a higher-priority bookmark or ability to jump to a specific page and easily configure this per document would be ideal here as it would allow bookmarks to be before and after the destination page(s) in question, but that's for another day. 

Thank you!

**Comment by hius07** [comment]
Go to first/last bookmark: https://github.com/koreader/koreader/pull/10899.
